### PR TITLE
create bot class

### PR DESCRIPTION
**Pull request by michaelenglo**
# Pull Request

## Description
Implement Bot class to interface with Slack

## Testing

If testing this change requires extra setup, please document it here:

## Ticket(s)

Affects #7 #27 

(Create a copy of that line for each Github Issue affected,
and replace "Affects" with "Closes" if merging this will close the relevant ticket.)

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -5,12 +5,14 @@ name = "pypi"
...
 [packages]
 pytest = "*"
+slackclient = "*"
 
 [dev-packages]
 pycodestyle = "*"
 codecov = "*"
 pytest-cov = "*"
 pydocstyle = "*"
+pylint = "*"
```
Comment (rwblickhan) [inline Pipfile]: Why did you add `pylint`?

File: bot/bot.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import os
```
Comment (rwblickhan) [inline bot/bot.py]: Please add [docstrings](https://www.python.org/dev/peps/pep-0257/) for this module and its functions. (This is causing the Travis build to fail at the moment.)

File: bot/bot.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+from slackclient import SlackClient
+
+slack_token = os.environ["SLACK_BOT_TOKEN"]
+
+class Bot:
+
+  def __init__(self):
+    self.sc = SlackClient(slack_token)
+
+  def sendMsg(self, message, channel):
```
Comment (rwblickhan) [inline bot/bot.py]: Can you add unit tests for this method? You will probably want to use [`unittest.mock`](https://docs.python.org/3/library/unittest.mock.html) to mock the `SlackClient` (let me know if you want some help looking at that - it may require you to refactor the class a bit to follow [dependency inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle) better).

File: bot/bot.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+"""Utility class for interacting with Slack API"""
+import os
+from slackclient import SlackClient
+
+slack_token = os.environ["SLACK_API_TOKEN"]
```
Comment (rwblickhan) [inline bot/bot.py]: It would probably be a good idea to move this into the `__init__` function for the actual bot class. (Also, we would want to make this configurable at some point, e.g. via a yaml config file, but that doesn't need to be in this pull request.)

File: bot/bot.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  def send_dm(self, message, slack_user_id):
+    """Send direct message to user with id of slack_user_id"""
+    response = self.sc.api_call(
+      "chat.postMessage",
+      channel=slack_user_id,
+      text=message
+    )
+
+    if not response['ok']:
+      raise Exception('API Call error: {}'.format(response["error"]))
```
Comment (rwblickhan) [inline bot/bot.py]: This is fine for now but I think in the long run raising exceptions will be somewhat problematic - once we set up logging we'll probably want this to log an error but otherwise return silently.

File: bot/bot.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    response = self.sc.api_call(
+      "chat.postMessage",
+      channel=channel_name,
+      text=message
+    )
+
+    if not response['ok']:
+      raise Exception('API Call error: {}'.format(response["error"]))
+
+    return
```
Comment (rwblickhan) [inline bot/bot.py]: You don't need an explicit `return` at the end of a method if you're not returning anything.

File: tests/bot/bot_test.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+"""Test Bot Class."""
+from bot.bot import Bot
+
+def test_send_dm():
+    """Test the Bot class method send_dm()."""
+    bot = Bot()
```
Comment (rwblickhan) [inline tests/bot/bot_test.py]: These unit tests should probably be using mocks, so that we're not sending messages every time we run the unit tests.

File: bot/bot.py (hunk 4 above)
Comment (cheukyin699) [reply bot/bot.py]: Also try `os.environ.get('SLACK_API_TOKEN', <default value>)` to prevent key errors.

**Comment by michaelenglo** [comment]
@rwblickhan I am not sure how to best resolve the Pipfile.lock conflict here. Any thoughts?

**Comment by rwblickhan** [comment]
This branch is out of date now, right? It’s probably best to bring it up to date with a rebase, though that would also require fixing the `Pipfile.lock` conflicts. Alternatively we could do some “git surgery” with cherry-picks and all, though we’d want to be careful about that. Maybe it’s best to discuss it this weekend.
